psi4/src/psi4/libfock/jk.h

```diff
@@ -1191,6 +1192,95 @@ class PSI_API MemDFJK : public JK {
     std::shared_ptr<DFHelper> dfh() { return dfh_; }
 };
 
+/**
+ * Class DFJCOSK
+ *
+ * JK implementation using a direct density-fitted coulomb algorithm
+ * and a semi-numerical 'chain of spheres' exchange algorithm
+ */
+class PSI_API DFJCOSK : public JK {
+   protected:
+    /// Number of threads
+    int nthreads_;
+    /// Options object
+    Options& options_;
+    /// Previous iteration pseudo-density matrix
+    std::vector<SharedMatrix> D_prev_;
+
+    // => Density Fitting Stuff <= //
+
+    /// Auxiliary basis set
+    std::shared_ptr<BasisSet> auxiliary_;
+    /// Coulomb Metric
+    SharedMatrix J_metric_;
+    /// per-thread TwoBodyAOInt object (for computing three-center ERIs)
+    std::vector<std::shared_ptr<TwoBodyAOInt>> eri_computers_;
+
+    // => Semi-Numerical Stuff <= //
+
+    /// Small DFTGrid for initial SCF iterations
+    std::shared_ptr<DFTGrid> grid_init_;
+    /// Large DFTGrid for the final SCF iteration
+    std::shared_ptr<DFTGrid> grid_final_;
+    /// Overlap fitting metric for grid_initial_
+    SharedMatrix Q_init_;
+    /// Overlap fitting metric for grid_final_
+    SharedMatrix Q_final_;
```

Comment:
None of these truly need to be a pointer. The one thing you need to do is change the definition of `grid` and `Q` to be ternaries, and then you can change the signature of `compute_numeric_overlap` to remove an annoying pointer.

Reply:
Were we to make `grid_init_`, `grid_final_`, `Q_init_`, and `Q_final_` objects instead of pointers, how would you initialize those objects without incurring a copy?

Reply:
If you mean "initialize `grid_init_`", you initialize it when you construct a new object. The use of pointers isn't saving you anything.
If you mean "initialize `grid`", then `const& grid = early_screening_ ? grid_init_ : grid_final_` should work.

Reply:
I'm talking about the former.

When `grid_init_` is a `shared_ptr<DFTGrid>`, I can write:
```
grid_init_ = std::make_shared<DFTGrid>(...);
```

But when `grid_init_` is a `DFTGrid`, I have to write:
```
grid_init_ = DFTGrid(...);
```
which from my understanding first creates a grid and then assigns it to `grid_init_` (requiring a copy). RVO doesn't apply here, because `grid_init_` is already defined; RVO only works for initialization.

Reply:
[Copy-assignment, not copy](https://stackoverflow.com/questions/37975574/what-happens-when-a-new-object-is-created-using-another-existing-object), but your point stands. (And I have learned something new today.) _Technically_ it's more efficient to use a `unique_ptr` than a `shared_ptr`, but I can't imagine that's significant for this application.